### PR TITLE
#2104 - Integrate Roles Permission Tab Functionality Into Customization (D.T) Beta

### PR DESCRIPTION
**Pull request by kodinkat**
- fixes: #2104 
---
 
Screenshot 2023-07-05 at 09 04 37

### CODE REVIEW

**Comment by corsacca** [comment]
Weird double view.
image


Probably `ha` and `haha` ?

**Comment by corsacca** [comment]
Nice work on this one @kodinkat!

In the UI and in the API, only allow updating the administrator role if the current using in an administrator (use `dt_is_administrator()` function).

**Comment by corsacca** [comment]
Did you see an easy way to add the rest of the contact caps to the list too?

**Comment by kodinkat** [comment]
> Weird double view. image
> 
> Probably `ha` and `haha` ?

Unable to recreate... Hopefully latest changes have resolved glitch..

**Comment by kodinkat** [comment]
> Did you see an easy way to add the rest of the contact caps to the list too? ( Each post type listing their own custom capabilities)

`dt_custom_post_types` option? However, seems to list the default 5 capabilities defined within `dt_capabilities`.

Also, are you wanting to add WordPress capabilities under each role?

Screenshot 2023-07-13 at 12 49 33

File: dt-core/admin/admin-settings-endpoints.php [hunk 1]
```diff
@@ -222,6 +264,172 @@ public static function get_post_fields() {
...
+            $custom_post_types = get_option( 'dt_custom_post_types', [] );
+
+            // Proceed with storing updates, overwriting existing roles.
+            $post_type_settings = !empty( $custom_post_types[$post_type] ) ? $custom_post_types[$post_type] : [];
+            $post_type_settings['roles'] = [];
+
+            foreach ( $roles as $role => $capabilities ){
+                foreach ( $capabilities ?? [] as $capability ){
+                    if ( $capability['enabled'] ){
+                        if ( ( $role == 'administrator' && dt_is_administrator() ) || ( $role != 'administrator' ) ){
```
Comment (kodinkat) [inline dt-core/admin/admin-settings-endpoints.php:415]: @corsacca

File: dt-core/admin/menu/tabs/tab-customizations.php [hunk 2]
```diff
@@ -333,6 +346,165 @@ private function get_post_fields( $post_type ){
...
+                                        $capability_name = $capability->name ?? $capability_key;
+                                        ?>
+                                        <div class="capability">
+                                            <label>
+                                                <input  type="checkbox"
+                                                        class="roles-settings-capability"
+                                                        data-role_key="<?php echo esc_attr( $key ) ?>"
+                                                        data-capability_key="<?php echo esc_attr( $capability_key ) ?>"
+                                                        <?php echo esc_attr( $is_capability_selected ? 'checked' : '' ) ?>
+                                                        <?php echo esc_attr( ( $key == 'administrator' ) && !dt_is_administrator() ? 'disabled' : '' ) ?>
```
Comment (kodinkat) [inline dt-core/admin/menu/tabs/tab-customizations.php:386]: @corsacca

File: dt-core/admin/multi-role/inc/class-capability-factory.php [hunk 3]
```diff
@@ -161,8 +161,17 @@ private function restricted_capabilities() {
      */
     public function add_capability( $slug, $options ) {
         $source = $options['source'];
```
Comment (kodinkat) [inline dt-core/admin/multi-role/inc/class-capability-factory.php:163]: @corsacca

**Pull request merged**